src/TestFramework/AbstractTestFrameworkAdapter.php

```diff
@@ -123,34 +124,19 @@ public function getMutantCommandLine(
         );
     }
 
+    /**
+     * @throws UnsupportedTestFrameworkVersion
+     */
     public function getVersion(): string
     {
-        if ($this->version !== null) {
-            return $this->version;
-        }
-
-        $testFrameworkVersionExecutable = $this->commandLineBuilder->build(
-            $this->testFrameworkExecutable,
-            [],
-            ['--version']
-        );
-
-        $process = new Process($testFrameworkVersionExecutable);
-        $process->mustRun();
-
-        $version = 'unknown';
-
-        try {
-            $version = $this->versionParser->parse($process->getOutput());
-        } catch (InvalidArgumentException $e) {
-            $version = 'unknown';
-        } finally {
-            $this->version = $version;
-        }
-
-        return $this->version;
+        return $this->version ?? $this->version = $this->retrieveVersion();
     }
 
+    /**
+     * @throws UnsupportedTestFrameworkVersion
+     */
+    abstract public function checkVersion(): void;
```

Comment:
this should be pushed in the `TestFrameworkAdapter` interface. The goal IMO is for it to be called by the test framework factory (not the adapter specific one, but the `Factory` one like done in this PR)

Reply:
1. note that this class in not used by any other adapter
2. we need to make it inside `abstract-testframework-adapter` first, not here

Reply:
yes, but that was a PoC :)